tests/std/tests/P0896R4_ranges_algorithm_machinery/test.cpp

```diff
@@ -533,3 +533,91 @@ namespace result_test {
+            proxy& operator=(proxy const&) = delete;
+
+            operator int() const;
+            // 1: not reflexively indirectly_movable_storable
+            void operator=(int) const requires(I != 1);
+        };
+
+        proxy operator*() const;
+
+        archetype& operator++();
+        archetype operator++(int);
+
+        // 0: not forward_iterator (input only)
+        bool operator==(archetype const&) const requires(I != 0) = default;
+
+        friend int iter_move(archetype const&) {
+            return 42;
+        }
+        friend void iter_swap(archetype const&, archetype const&) {}
+
+        // Ideally we'd have a "not reflexively indirectly_swappable" case as well, but there's no way for a class to
+        // satisfy indirectly_movable_storable<T, T> without satisfying indirectly_swappable<T, T> thanks to N4861
+        // [iterator.cust.swap]/4.3. permutable requires indirectly_swappable<T, T> only to forbid a user type to define
+        // an iter_swap overload that doesn't meet the semantic requirements.
+    };
+    STATIC_ASSERT(std::input_iterator<archetype<0>>);
+    STATIC_ASSERT(!std::forward_iterator<archetype<0>>);
+    STATIC_ASSERT(indirectly_movable_storable<archetype<0>, archetype<0>>);
+    STATIC_ASSERT(indirectly_swappable<archetype<0>, archetype<0>>);
+
+    STATIC_ASSERT(std::forward_iterator<archetype<1>>);
+    STATIC_ASSERT(!indirectly_movable_storable<archetype<1>, archetype<1>>);
+    STATIC_ASSERT(indirectly_swappable<archetype<1>, archetype<1>>);
+
+    STATIC_ASSERT(std::forward_iterator<archetype<2>>);
+    STATIC_ASSERT(indirectly_movable_storable<archetype<2>, archetype<2>>);
+    STATIC_ASSERT(indirectly_swappable<archetype<2>, archetype<2>>);
+
+    STATIC_ASSERT(!permutable<archetype<0>>);
+    STATIC_ASSERT(!permutable<archetype<1>>);
+    STATIC_ASSERT(permutable<archetype<2>>);
+} // namespace permutable_test
+
+namespace sortable_test {
+    using ranges::less;
+    using std::identity, std::indirect_strict_weak_order, std::permutable, std::projected, std::sortable;
+
+    void test() {
+        {
+            using I = int const*; // not permutable
+            using C = less;
+            using P = identity;
+            STATIC_ASSERT(!permutable<I>);
+            STATIC_ASSERT(indirect_strict_weak_order<C, projected<I, P>>);
+            STATIC_ASSERT(!sortable<I, C, P>);
+        }
+
+        {
+            using I = int*;
+            using C = void; // not an indirect_strict_weak_order
+            using P = identity;
+            STATIC_ASSERT(permutable<I>);
+            STATIC_ASSERT(!indirect_strict_weak_order<C, projected<I, P>>);
+            STATIC_ASSERT(!sortable<I, C, P>);
+        }
+
+        {
+            using I = int*;
+            using C = less;
+            using P = identity;
+            STATIC_ASSERT(permutable<I>);
+            STATIC_ASSERT(indirect_strict_weak_order<C, projected<I, P>>);
+            STATIC_ASSERT(sortable<I, C, P>);
+        }

```

Comment:
Can/should there be test coverage for a projection that interferes with sortability? Currently, only `identity` is tested. Alternatively, there could be test coverage for a type that is permutable, but not comparable via `identity`, only when viewed through the lens of a projection.

Reply:
We cover all variations of properties that can cause a set of arguments to not satisfy `indirect_strict_weak_order` in the test for `indirect_strict_weak_order`. Here, `indirect_strict_weak_order` and `permutable` are opaque - we only care that `sortable` is satisfied when both of them are satisfied, and that `sortable` is not satisfied if either of them is not satisfied.

This is exactly the method of testing we've been using for other concepts.